txncache: improvements

yufeng-jump opened the pull request:
- Support worst case transaction mix under mainnet conditions, including nonce transactions.
- Implement race free reprobing.
- Adjust struct layout for better cache line behavior: up to 43% speedup for queries.
- Make the size of txncache configurable.
- Add support for multiple key sizes.
- Remove support for serializing the txncache into a snapshot.
- Better support for key slicing after loading from an Agave snapshot.
- Remove redundant read lock acquisition in executor root slot check.
- More tests, including one that tries to simulate production op mixes.
- Add metrics.

mmcgee-jump reviewed (changes requested):
Just flagging for review, will take a look soon

cali-jumptrading commented on src/flamenco/runtime/fd_txncache.c:
```diff
@@ -983,19 +1129,28 @@ fd_txncache_query_batch( fd_txncache_t *             tc,
-    ulong head_hash = FD_LOAD( ulong, query->txnhash+txnhash_offset ) % FD_TXNCACHE_BLOCKCACHE_MAP_CNT;
-    for( uint head=blockcache->heads[ head_hash ]; head!=UINT_MAX; head=txnpages[ head/FD_TXNCACHE_TXNS_PER_PAGE ].txns[ head%FD_TXNCACHE_TXNS_PER_PAGE ]->blockcache_next ) {
+    if( FD_UNLIKELY( txnhash_offset) ) {
+      ulong max_key_idx = fd_ulong_sat_sub( query->key_sz, FD_TXNCACHE_KEY_SIZE+1UL );
```
maybe some sanity checks on key_sz. Or... if key size is inherently a property of the `txncache`, then maybe it doesn't make sense for the user to supply the key size when calling `txncache` APIs. Rather, the user could supply the key sizes for txn hash and block hash on `txncache` creation.

cali-jumptrading reviewed (commented):
Was mostly reviewing this PR to learn more about the status cache

topointon-jump commented on src/flamenco/runtime/fd_executor.c:
```diff
@@ -259,20 +259,38 @@ fd_validate_fee_payer( fd_txn_account_t * account,
 
 static int
 status_check_tower( ulong slot, void * _ctx ) {
```
Note for future follow-up - after the runtime re-architecture, it might be worth exploring if we can make the status cache a serial data structure which is only accessed from a single tile, and remove all these locks and atomic operations.

topointon-jump commented on src/flamenco/runtime/fd_txncache.c:
```diff
@@ -28,44 +27,62 @@
-  to the entry, but stop iterating while running queries. */
+/* The number of unique entries in the hash lookup table for each
+   (slot) in the slotcache.
+   FIXME This is going to fill up fast if there are lots of nonces in a slot
```
We need to address this - can do this in a follow-up PR

topointon-jump commented on src/flamenco/runtime/fd_txncache.c:
```diff
@@ -969,33 +1127,50 @@ fd_txncache_query_batch( fd_txncache_t *             tc,
+  FD_ATOMIC_FETCH_AND_ADD( &tc->query_absent,  query_absent  );
 }
 
 int
```
We should delete `fd_txncache_snapshot` and all the accompanying code/data structures, this is never used and `fd_txncache_get_entries` is used instead.

yufeng-jump replied:
Deleted.

yufeng-jump commented on src/flamenco/runtime/fd_txncache.c:
```diff
@@ -469,38 +500,122 @@ fd_txncache_delete( void * shtc ) {
-      blockcache[ j ].max_slot = FD_TXNCACHE_EMPTY_ENTRY;
-    }
-    j = (j+1)%tc->live_slots_max;
+static void
+fd_txncache_print_insert( fd_txncache_insert_t const * txn,
+                          char const *                 prefix ) {
+
+  FD_BASE58_ENCODE_32_BYTES( txn->blockhash, blockhash );
+  if( txn->key_sz==32UL ) {
+    FD_BASE58_ENCODE_32_BYTES( txn->txnhash, key_str );
+    FD_LOG_NOTICE(( "%s: blockhash %s, slot %lu, key %s, res %u",
+                    prefix,
+                    blockhash,
+                    txn->slot,
+                    key_str,
+                    *(txn->result) ));
+  } else if( txn->key_sz==64UL ) {
+    FD_BASE58_ENCODE_64_BYTES( txn->txnhash, key_str );
+    FD_LOG_NOTICE(( "%s: blockhash %s, slot %lu, key %s, res %u",
+                    prefix,
+                    blockhash,
+                    txn->slot,
+                    key_str,
+                    *(txn->result) ));
+  } else {
+    FD_LOG_NOTICE(( "%s: blockhash %s, slot %lu, key %s, res %u",
+                    prefix,
+                    blockhash,
+                    txn->slot,
+                    "(unsupported key length)",
+                    *(txn->result) ));
   }
+}
```
This is basically an irrecoverable error.  The validator is gonna crash if it can't insert.  So I think NOTICE is fine here.

yufeng-jump commented on src/flamenco/runtime/fd_txncache.c:
```diff
@@ -553,20 +672,32 @@ static void
 
   if( FD_UNLIKELY( tc->root_slots_cnt>=tc->root_slots_max ) ) {
     if( FD_LIKELY( idx ) ) {
+      long start = fd_log_wallclock();
       fd_txncache_purge_slot( tc, root_slots[ 0 ] );
+      long end = fd_log_wallclock();
+      FD_LOG_INFO(( "registering slot %lu at idx %lu triggered a purge of %ld nanos", slot, idx, end-start ));
```
Purging happens very infrequently.  We do wallclock calls even in pack, and they aren't all that expensive at the frequency we call it.  I can see the argument for doing a metrics histogram rather than a log info here.  That cleanup can be a followup PR.

yufeng-jump committed:
txncache: faster higher stronger

- Support worst case transaction mix under mainnet conditions, including
  nonce transactions.

- Implement race free reprobing.

- Adjust struct layout for better cache line behavior: up to 43% speedup
  for queries.

- Make the size of txncache configurable.

- Add support for multiple key sizes.

- Remove support for serializing the txncache into a snapshot.

- Better support for key slicing after loading from an Agave snapshot.

- Remove redundant read lock acquisition in executor root slot check.

- More tests, including some that try to simulate production op mixes.

- Add metrics.
